### PR TITLE
Add `post_shutdown`

### LINKED ISSUE #3105: post_shutdown method in addition to post_init
### What kind of feature are you missing? Where do you notice a shortcoming of PTB?

i would like to be able to await coroutines before of shutting down the bot after application.run_polling().

it could be done implementing a post_shutdown method in addition to the already existing post_init method

### Describe the solution you'd like

It is especially useful for closing a db connection in aiosqlite ( await db.close() )

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ExalFabu**
This should resolve #3105 
I strictly followed #3078, I hope I've not missed any documentation that was required to change as well

Anyway, as I mentioned in the issue i am struggling with these test that i added:  
- `test_run_polling_post_shutdown`
- `test_run_webhook_post_shutdown`

Here's a screenshot of the error (it's the same for both tests)
image

I don't know if I've messed up the test itself or I'm missing something

### CODE REVIEW

File: telegram/ext/_applicationbuilder.py [hunk 1]
```diff
@@ -928,6 +931,42 @@ async def post_init(application: Application) -> None:
...
+
+        Tip:
+            This can be used for custom shutdown logic that requires to await coroutines, e.g.
+            closing a database connection
+
+        Example:
+            .. code::
+
+                async def post_shutdown(application: Application) -> None:
+                    await db_connection.close()
```
Comment (Bibo-Joshi) [inline telegram/ext/_applicationbuilder.py:950]:
```suggestion
                    await application.bot_data['database'].close()
```

just an idea - makes it more obvious why we might need the `application` parameter

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the PR! I left two comments below

**Comment by ExalFabu** [comment]
I fixed and removed that duplicated line of code and the tests i wrote passed.
Though it seems another test broke: `test_run_error_in_application`, only when the error is simulated inside `application.initialize()`

And for what I understand it's really not an issue, since at this point of the execution the updater hasn't been initialized yet, and that's why `app.shutdown()` doesn't call `updater.shutdown()`

If my assumptions are correct, i may fix the test asserting `app._initialized` and `app.updater._initilized` as false, removing the need to monkeypatch the methods
Alternatively i can assert the two different cases inside a if statement
Something like `assert set(shutdowns) == ( {"application", "updater"} if method == "start" else {"application"} ) `

**Comment by Bibo-Joshi** [comment]
The bad boy is this here:

https://github.com/python-telegram-bot/python-telegram-bot/blob/f1d03393def74bfc09f48556ea2b20e8c6098189/telegram/ext/_application.py#L374-L376

this tells `Application.shutdown` to just exit if the app wasn't properly initialized and hence `Updater.shutdown` is not called.
While we can't be 100% sure that `Updater.initialize` wasn't called by `Application.initialize`, I'm okay with that. Catching all errors that can happen during startup is a kind of a sisyphus task, so a best effort suffices IMO.

File: telegram/ext/_applicationbuilder.py [hunk 2]
```diff
@@ -928,6 +931,42 @@ async def post_init(application: Application) -> None:
         self._post_init = post_init
         return self
 
+    def post_shutdown(
+        self: BuilderType, post_shutdown: Callable[[Application], Coroutine[Any, Any, None]]
+    ) -> BuilderType:
+        """
+        Sets a callback to be executed by :meth:`Application.run_polling` and
+        :meth:`Application.run_webhook` *after* executing :meth:`Application.shutdown` but
+        *before* executing :meth:`Updater.shutdown` respectively.
```
Comment (harshil21) [inline telegram/ext/_applicationbuilder.py:940]:
```suggestion
        :meth:`Application.run_webhook` *after* executing :meth:`Updater.shutdown`
        and :meth:`Application.shutdown`.
```

**Review by harshil21 — COMMENTED** [review]
just one comment

File: telegram/ext/_applicationbuilder.py (hunk 2 above)
Comment (ExalFabu) [reply telegram/ext/_applicationbuilder.py:940]: Oops, missed that comment

Is this up to me to commit? If yes let me know, i don't want to hold this PR just for this (sorry)

Comment (harshil21) [reply telegram/ext/_applicationbuilder.py:940]: yeah, you can just commit my suggestion (click the button above)

**Review by harshil21 — APPROVED** [review]

**Comment by Bibo-Joshi** [comment]
Thank you for the contribution :)

**Pull request merged**